data-prepper-plugins/http-source/src/test/java/com/amazon/dataprepper/plugins/source/loghttp/HTTPSourceTest.java

```diff
@@ -96,10 +101,51 @@ public void testHTTPJsonResponse() {
                         .build(),
                 HttpData.ofUtf8("[{\"log\": \"somelog\"}]"))
                 .aggregate()
-                .whenComplete((i, ex) -> assertThat(i.status().code()).isEqualTo(200)).join();
+                .whenComplete((i, ex) -> assertThat(i.status()).isEqualTo(HttpStatus.OK)).join();
         Assertions.assertFalse(testBuffer.isEmpty());
     }
 
+    @Test
+    public void testHTTPJsonResponse429() {
+        // Prepare
+        final Map<String, Object> settings = new HashMap<>();
+        final int testMaxPendingRequests = 1;
+        final int testThreadCount = 1;
+        settings.put(HTTPSourceConfig.MAX_PENDING_REQUESTS, testMaxPendingRequests);
+        settings.put(HTTPSourceConfig.THREAD_COUNT, testThreadCount);
+        testPluginSetting = new PluginSetting("http", settings);
+        testPluginSetting.setPipelineName("pipeline");
+        HTTPSourceUnderTest = new HTTPSource(testPluginSetting);
+        // Start the source
+        HTTPSourceUnderTest.start(testBuffer);
+        final RequestHeaders testRequestHeaders = RequestHeaders.builder().scheme(SessionProtocol.HTTP)
+                .authority("127.0.0.1:2021")
+                .method(HttpMethod.POST)
+                .path("/log/ingest")
+                .contentType(MediaType.JSON_UTF_8)
+                .build();
+        final HttpData testHttpData = HttpData.ofUtf8("[{\"log\": \"somelog\"}]");
+
+        // Fill in the buffer
+        WebClient.of().execute(testRequestHeaders, testHttpData).aggregate().whenComplete(
+                        (response, ex) -> assertThat(response.status()).isEqualTo(HttpStatus.OK)).join();
+
+        // Send requests to throttle the server when buffer is full
+        // Set the client timeout to be less than source request_timeout / (testMaxPendingRequests + testThreadCount)
+        WebClient testWebClient = WebClient.builder().responseTimeout(Duration.ofMillis(100)).build();
+        for (int i = 0; i < testMaxPendingRequests + testThreadCount; i++) {
+            try {
+                testWebClient.execute(testRequestHeaders, testHttpData).aggregate().join();
+            } catch (CompletionException e) {
+                assertThat(e.getCause()).isInstanceOf(ResponseTimeoutException.class);
+            }
+        }
+
+        // When/Then
+        testWebClient.execute(testRequestHeaders, testHttpData).aggregate().whenComplete(
+                        (response, ex) -> assertThat(response.status()).isEqualTo(HttpStatus.TOO_MANY_REQUESTS)).join();
+    }
```

Comment:
It would be good to also be sure that the server gets out of the 429 state after the requests clear up.